src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Claims/WindowsClaimSet.cs

```diff
@@ -122,29 +175,37 @@ public void Dispose()
             if (!_disposed)
             {
                 _disposed = true;
-                _windowsIdentity.Dispose();
+                if(_windowsIdentity is WindowsIdentity)
+                    ((WindowsIdentity)_windowsIdentity).Dispose();
             }
         }
 
         private IList<Claim> InitializeClaimsCore()
         {
-            if (_windowsIdentity.AccessToken == null)
+            List<Claim> claims = new List<Claim>();
+            if (_windowsIdentity is WindowsIdentity)
             {
-                return new List<Claim>();
-            }
+                WindowsIdentity _windowsInternalIdentity = (WindowsIdentity)_windowsIdentity;
+                if (_windowsInternalIdentity.AccessToken == null)
+                {
+                    return new List<Claim>();
+                }
 
-            List<Claim> claims = new List<Claim>(3)
-            {
-                new Claim(ClaimTypes.Sid, _windowsIdentity.User, Rights.Identity)
-            };
-            if (TryCreateWindowsSidClaim(_windowsIdentity, out Claim claim))
+                claims.Add(new Claim(ClaimTypes.Sid, _windowsInternalIdentity.User, Rights.Identity));
+                if (TryCreateWindowsSidClaim(_windowsInternalIdentity, out Claim claim))
+                {
+                    claims.Add(claim);
+                }
+                claims.Add(Claim.CreateNameClaim(_windowsIdentity.Name));
+            }
+            else
             {
-                claims.Add(claim);
+                claims.Add(Claim.CreateNameClaim(_windowsIdentity.Name));
             }
```

Comment:
This doesn't need to be in an else block as you also do it at the end of the true if block. Just move this outside of the if block and don't have an else.